src/script/sign.h

```diff
@@ -68,6 +71,432 @@ struct SignatureData {
+                }
+
+                ++separators;
+
+                continue;
+            }
+
+            // Read key
+            std::vector<unsigned char> key;
+            key.resize(key_len);
+            Span<unsigned char> key_data = MakeSpan(key);
+            s >> key_data;
+
+            // First byte of key is the type
+            unsigned char type = key[0];
+
+            // Read in value length
+            uint64_t value_len = ReadCompactSize(s);
+
+            // Do stuff based on type
+            switch(type) {
+                // Raw transaction or a non witness utxo
+                case PSBT_UNSIGNED_TX_NON_WITNESS_UTXO:
+                    if (in_globals) {
+                        s >> tx;
+                    } else {
+                        // Read in the transaction
+                        CTransactionRef prev_tx;
+                        s >> prev_tx;
+
+                        // Check that this utxo matches this input
+                        if (tx.vin[input.index].prevout.hash.Compare(prev_tx->GetHash()) != 0) {
+                            throw std::ios_base::failure("Provided non witness utxo does not match the required utxo for input");
+                        }
+
+                        // Add to input
+                        input.non_witness_utxo = std::move(prev_tx);
+                    }
+                    break;
+                // redeemscript or a witness utxo
+                case PSBT_REDEEMSCRIPT_WITNESS_UTXO:
+                    if (in_globals) {
+                        // Make sure that the key is the size of uint160 + 1
+                        if (key.size() != 21) {
+                            throw std::ios_base::failure("Size of key was not the expected size for the type redeem script");
+                        }
+                        // Retrieve hash160 from key
+                        uint160 hash160(std::vector<unsigned char>(key.begin() + 1, key.end()));
+
+                        // Read in the redeemscript
+                        std::vector<unsigned char> redeemscript_bytes;
+                        redeemscript_bytes.resize(value_len);
+                        Span<unsigned char> redeemscript_data = MakeSpan(redeemscript_bytes);
+                        s >> redeemscript_data;
+                        CScript redeemscript(redeemscript_bytes.begin(), redeemscript_bytes.end());
+
+                        // Check that the redeemscript hash160 matches the one provided
+                        hash160_hasher.Write(&redeemscript_bytes[0], redeemscript_bytes.size());
+                        unsigned char rs_hash160_data[hash160_hasher.OUTPUT_SIZE];
+                        hash160_hasher.Finalize(rs_hash160_data);
+                        uint160 rs_hash160(std::vector<unsigned char>(rs_hash160_data, rs_hash160_data + hash160_hasher.OUTPUT_SIZE));
+                        if (hash160.Compare(rs_hash160) != 0) {
```

Comment:
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

Doesn't really make a huge difference, but I think `if (hash160 != rs_hash160)` is a tiny bit more readable.

Reply:
Done